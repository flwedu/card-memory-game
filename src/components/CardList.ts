import Card from "./Card";
import Component from "./Component";

export default class CardList implements Component {

    cards: Card[] = [];

    constructor(private el: HTMLElement, private numberArray: number[]) {
        numberArray.forEach(
            value => this.addCard(value)
        )
    }

    addCard(value: number) {
        // Creating entity
        const card = new Card(value);
        this.cards.push(card);
        // Creating child node
        this.el.appendChild(card.getHtmlElement());
    }

    render(): void {
    }

}