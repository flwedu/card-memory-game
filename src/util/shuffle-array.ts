export function shuffle(array: any[]) {
    let lastIndex = array.length, randomIndex;

    while (lastIndex != 0) {

        randomIndex = Math.floor(Math.random() * lastIndex);
        lastIndex--;

        [array[lastIndex], array[randomIndex]] = [array[randomIndex], array[lastIndex]];
    }

    return array;
}