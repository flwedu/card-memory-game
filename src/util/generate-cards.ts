import Card from "../components/Card";
import { shuffleArray } from "./shuffle-array";

export function generateCardsArrayWithDoubleLength(length: number) {
  const numberArray: number[] = [];

  for (let i = 0; i < length; i++) {
    numberArray.push(i + 1);
    numberArray.push(i + 1);
  }

  const shuffledArray = shuffleArray(numberArray);

  const cards = generateCardsArrayFromArray(shuffledArray);

  return cards;
}

export function generateCardsArrayFromArray(array: number[]) {
  const cards = array.map((value, index) => {
    return new Card(value, index);
  });

  return cards;
}
