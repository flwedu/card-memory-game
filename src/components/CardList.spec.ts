import { generateCardsArrayFromArray } from "../util/generate-cards";
import CardList from "./CardList";

describe("CardList class tests", () => {
  const numberArr = [1, 1, 2, 2];
  test("render() should return the correct html to div element", () => {
    const div = document.createElement("div");
    const cards = generateCardsArrayFromArray(numberArr);
    const cardList = new CardList(div, cards);

    expect(div.innerHTML).toMatchSnapshot();
    expect(cardList.render()).toEqual(div.innerHTML);
  });

  test("handleClick() should call the flip() method of the card", () => {
    const div = document.createElement("div");
    const cards = generateCardsArrayFromArray(numberArr);
    const cardList = new CardList(div, cards);
    const firstCard = cards[0];
    const spy = jest.spyOn(firstCard, "flip");

    const firstCardEl = firstCard.getEl();
    firstCardEl.dispatchEvent(new MouseEvent("click", { bubbles: true }));

    expect(spy).toHaveBeenCalled();
    expect(firstCard.flipped).toBe(true);
  });

  test("getSelectedCards() should return the selected cards", () => {
    const div = document.createElement("div");
    const cards = generateCardsArrayFromArray(numberArr);
    const cardList = new CardList(div, cards);
    const firstCard = cards[0];
    const secondCard = cards[1];

    const firstCardEl = firstCard.getEl();
    const secondCardEl = secondCard.getEl();

    expect(cardList.getSelectedCards()).toEqual([]);

    firstCardEl.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    secondCardEl.dispatchEvent(new MouseEvent("click", { bubbles: true }));

    expect(cardList.getSelectedCards()).toEqual([firstCard, secondCard]);
  });

  describe("checkSelectedCardsMatch()", () => {
    test("should return true if the selected cards match, and getFlippedCards() should return 2 cards", () => {
      const numberArr = [1, 1];
      const div = document.createElement("div");
      const cards = generateCardsArrayFromArray(numberArr);
      const cardList = new CardList(div, cards);
      const firstCard = cards[0];
      const secondCard = cards[1];

      const firstCardEl = firstCard.getEl();
      const secondCardEl = secondCard.getEl();

      firstCardEl.dispatchEvent(new MouseEvent("click", { bubbles: true }));
      secondCardEl.dispatchEvent(new MouseEvent("click", { bubbles: true }));

      expect(cardList.checkSelectedCardsMatch()).toBe(true);
      expect(cardList.getFlippedCards().length).toBe(2);
      expect(cardList.getUnmatchedCards().length).toBe(0);
    });

    test("should return false if the selected cards don't match, and getFlippedCards() should return 0 cards", () => {
      const numberArr = [1, 2];
      const div = document.createElement("div");
      const cards = generateCardsArrayFromArray(numberArr);
      const cardList = new CardList(div, cards);
      const firstCard = cards[0];
      const secondCard = cards[1];

      const firstCardEl = firstCard.getEl();
      const secondCardEl = secondCard.getEl();

      firstCardEl.dispatchEvent(new MouseEvent("click", { bubbles: true }));
      secondCardEl.dispatchEvent(new MouseEvent("click", { bubbles: true }));

      expect(cardList.checkSelectedCardsMatch()).toBe(false);
      expect(cardList.getFlippedCards().length).toBe(0);
      expect(cardList.getUnmatchedCards().length).toBe(2);
    });
  });
});
