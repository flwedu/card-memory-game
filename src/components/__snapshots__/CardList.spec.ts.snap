// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardList class tests render() should return the correct html to div element 1`] = `"<div data-index=\\"0\\" class=\\"card\\"><img src=\\"assets/0.png\\"></div><div data-index=\\"1\\" class=\\"card\\"><img src=\\"assets/0.png\\"></div><div data-index=\\"2\\" class=\\"card\\"><img src=\\"assets/0.png\\"></div><div data-index=\\"3\\" class=\\"card\\"><img src=\\"assets/0.png\\"></div>"`;
